File: monabs/cores/dis_check.py
"""
Disjunctive over-approximation
"""
from typing import List
import z3


def compact_check_misc(precond: z3.ExprRef, cnt_list: [z3.ExprRef], res_label: List):
    f = z3.BoolVal(False)
    for i in range(len(res_label)):
        if res_label[i] == 2:
            f = z3.Or(f, cnt_list[i])
    if z3.is_false(f):
        return

    # sol = z3.SolverFor("QF_BV")
    sol = z3.Solver()
    g = z3.And(precond, f)
    sol.add(g)
    s_res = sol.check()
    if s_res == z3.unsat:
        for i in range(len(res_label)):
            if res_label[i] == 2:
                res_label[i] = 0
    elif s_res == z3.sat:
        m = sol.model()
        # models.append(m)  # counterexample
        for i in range(len(res_label)):
            if res_label[i] == 2 and z3.is_true(m.eval(cnt_list[i], True)):
                res_label[i] = 1
    else:
        return
    compact_check_misc(precond, cnt_list, res_label)


def disjunctive_check(precond: z3.ExprRef, cnt_list: List[z3.ExprRef]) -> List[int]:
    """
    Given a precond G and a set of cnts: f1, f2,..., fn
    Decis if the following cnts are satisfiable:
           And(G, f1), And(G, f2), ..., And(G, fn)
    Examples:
    >>> from z3 import *
    >>> x, y = Reals('x y')
    >>> pre = x > 100
    >>> f1 = x > 0; f2 = And(x > y, x < y); f3 = x < 3
    >>> cnts = [f1, f2, f3]
    >>> assert compact_check(pre, cnts) == [1, 0, 0]
    """
    res = []
    for _ in cnt_list:
        res.append(2)
    compact_check_misc(precond, cnt_list, res)
    return res

File: cpp/lib/dis_check.cpp
#include <z3++.h>
#include <vector>

void compact_check_misc(z3::expr precond, std::vector<z3::expr>& cnt_list, std::vector<int>& res_label) {
    z3::expr f = precond.ctx().bool_val(false);
    for (size_t i = 0; i < res_label.size(); ++i) {
        if (res_label[i] == 2) {
            f = z3::or_(f, cnt_list[i]);
        }
    }
    if (z3::eq(f, precond.ctx().bool_val(false))) {
        return;
    }

    z3::solver sol(precond.ctx());
    z3::expr g = z3::and_(precond, f);
    sol.add(g);
    auto s_res = sol.check();
    if (s_res == z3::unsat) {
        for (size_t i = 0; i < res_label.size(); ++i) {
            if (res_label[i] == 2) {
                res_label[i] = 0;
            }
        }
    } else if (s_res == z3::sat) {
        z3::model m = sol.get_model();
        for (size_t i = 0; i < res_label.size(); ++i) {
            if (res_label[i] == 2 && z3::eq(m.eval(cnt_list[i], true), precond.ctx().bool_val(true))) {
                res_label[i] = 1;
            }
        }
    } else {
        return;
    }
    compact_check_misc(precond, cnt_list, res_label);
}

std::vector<int> disjunctive_check(z3::expr precond, std::vector<z3::expr> cnt_list) {
    std::vector<int> res(cnt_list.size(), 2); // Initialize with 2 (unknown)
    compact_check_misc(precond, cnt_list, res);
    return res;
}

void compact_check_misc_incremental(z3::solver& solver, z3::expr precond, std::vector<z3::expr>& cnt_list, std::vector<int>& res_label) {
    z3::expr f = precond.ctx().bool_val(false);
    for (size_t i = 0; i < res_label.size(); ++i) {
        if (res_label[i] == 2) {
            f = z3::or_(f, cnt_list[i]);
        }
    }
    if (z3::eq(f, precond.ctx().bool_val(false))) {
        return;
    }

    solver.push();
    solver.add(f);
    auto s_res = solver.check();

    if (s_res == z3::unsat) {
        for (size_t i = 0; i < res_label.size(); ++i) {
            if (res_label[i] == 2) {
                res_label[i] = 0;
            }
        }
    } else if (s_res == z3::sat) {
        z3::model m = solver.get_model();
        for (size_t i = 0; i < res_label.size(); ++i) {
            if (res_label[i] == 2 && z3::eq(m.eval(cnt_list[i], true), precond.ctx().bool_val(true))) {
                res_label[i] = 1;
            }
        }
    }

    solver.pop();
    compact_check_misc_incremental(solver, precond, cnt_list, res_label);
}

std::vector<int> disjunctive_check_incremental(z3::expr precond, std::vector<z3::expr> cnt_list) {
    std::vector<int> res(cnt_list.size(), 2);
    z3::solver solver(precond.ctx());
    solver.add(precond);
    compact_check_misc_incremental(solver, precond, cnt_list, res);
    return res;
}
